Packs/ArmisEventCollector/Integrations/ArmisEventCollector/ArmisEventCollector.py

```diff
@@ -0,0 +1,427 @@
+
+    Args:
+        events (list[dict]): List of events from the current fetch response.
+        events_last_fetch_ids (list[str]): List of IDs of events from last fetch cycle.
+        unique_id_key (str): Unique event ID key of specific event type (Alert, Threat Activity etc.)
+
+    Returns:
+        tuple[list[dict], list[str]: The list of dedup events and ID list of events of current fetch.
+    """
+    # remove events that already appeared in last fetch
+    dedup_events: list[dict] = [event for event in events if event.get(unique_id_key) not in events_last_fetch_ids]
+    # get ids of dedup events
+    new_ids: list[str] = [event.get(unique_id_key, '') for event in dedup_events]
+
+    if not events:
+        return dedup_events, events_last_fetch_ids
+
+    latest_event_from_response = arg_to_datetime(events[-1].get('time'))
+    earliest_event_from_response = arg_to_datetime(events[0].get('time'))
+
+    if isinstance(latest_event_from_response, datetime) and isinstance(earliest_event_from_response, datetime)\
+            and are_two_event_time_equal(latest_event_from_response, earliest_event_from_response):
+        events_last_fetch_ids.extend(new_ids)
+        return dedup_events, events_last_fetch_ids
+    else:
+        return dedup_events, new_ids
+
+
+def fetch_by_event_type(event_type: dict, events: list, next_run: dict, client: Client,
+                        max_fetch: int, last_run: dict, fetch_start_time_param: datetime):
+    """ Fetch events by specific event type.
+
+    Args:
+        event_type (dict): Dictionary containing event type unique ID key, AQL query and type name.
+        events (list): List of fetched events.
+        next_run (dict): Last run dictionary for next fetch cycle.
+        client (Client): Armis client to use for API calls.
+        max_fetch (int): Max number of events to fetch.
+        last_run (dict): Last run dictionary.
+        fetch_start_time_param (datetime): Fetch start time (usually current time).
+    """
+    unique_id_key, aql_query, type = event_type['unique_id_key'], event_type['aql_query'], event_type['type']
+    last_fetch_ids = f'{type}_last_fetch_ids'
+    last_fetch_time = f'{type}_last_fetch_time'
+
+    demisto.debug(f'debug-log: handling event-type: {type}')
+    if last_run:
+        event_type_fetch_start_time = calculate_fetch_start_time(last_run.get(last_fetch_time), fetch_start_time_param)
+    else:
+        event_type_fetch_start_time = None
+
+    response = client.fetch_by_aql_query(
+        aql_query=aql_query,
+        max_fetch=max_fetch,
+        after=event_type_fetch_start_time
+    )
+    demisto.debug(f'debug-log: fetched {len(response)} {type} from API')
+    if response:
+        new_events, next_run[last_fetch_ids] = dedup_events(
+            response, last_run.get(last_fetch_ids, []), unique_id_key)
+        next_run[last_fetch_time] = new_events[-1].get('time') if new_events else last_run.get(last_fetch_time)
```

Comment:
Is it possible for dups between runs? If so we have to handle dups in between runs, by saving the ids of the last time (as exact as possible).
(fetched 1000 events at 12:00:01 but there were 2000 events in the same time)

Reply:
Yes.
Deduplication while paginating is not needed since the APIs pagination is index based and not time based.
For example: max_fetch is 200 but first page contains 100 events (events 0 to 99), so the 'next' attribute in the response will have the value of 100, which will be set in the 'from' paremter of the next request, letting the API know the exact index range of the next response (events 100 to 199).
As for deduplication betweeen fetch runs - there is saving of ID lists and last time for each event type, as well as handling the dedup usecases in dedup_events function (more details in dedup_events's docstring).